x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java

```diff
@@ -109,31 +109,28 @@ public void getUser(String username, ActionListener<User> listener) {
      */
     public void getUsers(String[] userNames, final ActionListener<Collection<User>> listener) {
         final Consumer<Exception> handleException = (t) -> {
-            if (t instanceof IndexNotFoundException) {
-                logger.trace("could not retrieve users because security index does not exist");
+            if (TransportActions.isShardNotAvailableException(t)) {
+                logger.trace("could not retrieve users because of a shard not available exception", t);
                 // We don't invoke the onFailure listener here, instead just pass an empty list
-                listener.onResponse(Collections.emptyList());
-            } else {
-                listener.onFailure(t);
             }
+            listener.onFailure(t);
         };
 
-        if (securityIndex.indexExists() == false) {
-            // TODO remove this short circuiting and fix tests that fail without this!
-            listener.onResponse(Collections.emptyList());
+        final SecurityIndexManager frozenSecurityIndex = this.securityIndex.freeze();
+        if (frozenSecurityIndex.isAvailable() == false) {
```

Comment:
Shouldn't we return an empty list if the index doesn't exist?

Otherwise we get
```
GET _xpack/security/user

{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index","index_uuid":"_na_","index":".security"}],"type":"index_not_found_exception","reason":"no such index","index_uuid":"_na_","index":".security"},"status":404}
```